src/decoder/builder.rs

```diff
@@ -263,29 +674,119 @@ impl<R: Read + Seek + Send + Sync + 'static> DecoderBuilder<R> {
     }
 
+    /// Provides a pre-computed total duration to avoid file scanning.
+    ///
+    /// When provided, decoders will use this value when they would otherwise need to scan the file.
+    ///
+    /// This affects decoder implementations that may scan for duration:
+    /// - **MP3**: May scan if metadata doesn't contain duration
+    /// - **Vorbis/OGG**: Scans to determine total duration
+    /// - **Symphonia**: Not controlled by this setting; may scan if byte_len is set
+    ///
+    /// # Examples
+    /// ```no_run
+    /// use std::time::Duration;
+    /// use std::fs::File;
+    /// use rodio::Decoder;
+    ///
+    /// let file = File::open("audio.mp3").unwrap();
+    /// let known_duration = Duration::from_secs(180);
+    ///
+    /// let decoder = Decoder::builder()
+    ///     .with_data(file)
+    ///     .with_total_duration(known_duration)  // Skip any scanning
+    ///     .build()
+    ///     .unwrap();
+    /// ```
+    pub fn with_total_duration(mut self, duration: Duration) -> Self {
+        self.settings.total_duration = Some(duration);
+        self
+    }
+
+    /// Enable file scanning for duration computation.
+    ///
+    /// **Important**: This setting only takes effect when the source is both seekable and has
+    /// `byte_len` is set. If these prerequisites are not met, this setting is ignored.
+    ///
+    /// This affects specific decoder implementations:
+    /// - **MP3**: May scan if metadata doesn't contain duration
+    /// - **Vorbis/OGG**: Scans to determine total duration
+    /// - **Symphonia**: Not controlled by this setting; may scan if byte_len is set
+    ///
+    /// File-based decoders (`Decoder::try_from(file)`) automatically enable this and set the
+    /// required prerequisites.
+    ///
+    /// # Examples
+    /// ```no_run
+    /// use std::fs::File;
+    /// use rodio::Decoder;
+    ///
+    /// let file = File::open("audio.mp3").unwrap();
+    /// let len = file.metadata().unwrap().len();
+    ///
+    /// # let file = std::fs::File::open("audio.mp3").unwrap();
+    /// # let len = file.metadata().unwrap().len();
+    /// let decoder = Decoder::builder()
+    ///     .with_data(file)
+    ///     .with_byte_len(len)        // Required
+    ///     .with_seekable(true)       // Already set by with_byte_len()
+    ///     .with_scan_duration(true)  // Now effective
+    ///     .build()
+    ///     .unwrap();
+    /// ```
+    pub fn with_scan_duration(mut self, scan: bool) -> Self {
+        self.settings.scan_duration = scan;
+        self
+    }
+
     /// Creates the decoder implementation with configured settings.
+    ///
+    /// This internal method handles the format detection and decoder creation process.
+    /// It attempts to create decoders in a specific order, passing the data source
+    /// between attempts until a compatible format is found.
+    ///
+    /// # Format Detection Order
+    ///
+    /// Decoders are tried in this order for optimal performance:
```

Comment:
I find the source code pretty readable, don't need the doc-comment. More afraid that the comment and code will get out of sync, and if they dont the presence of the doc makes it harder to refactor (as yo have to fix the docs too).